### PR TITLE
Weaviate core refactor

### PR DESCRIPTION
**Pull request by donomii**
### What's being changed:

This separates the http server from the weaviate code, allowing weaviate to be imported and run as a module.

It also adds accessors for various data structures so that they can be inspected.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/index.go [hunk 1]
```diff
@@ -1051,46 +1061,38 @@ func (i *Index) objectSearchByShard(ctx context.Context, limit int, filters *fil
...
-				}
-			} else {
-				objs, scores, err = i.remote.SearchShard(
-					ctx, shardName, nil, limit, filters, keywordRanking,
-					sort, cursor, nil, addlProps, i.replicationEnabled())
-				if err != nil {
-					return fmt.Errorf(
-						"remote shard object search %s: %w", shardName, err)
-				}
+		if shard := i.localShard(shardName); shard != nil {
```
Comment (antas-marcin) [inline adapters/repos/db/index.go:1072]: why did you remove running this method in goroutine?

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 2]
```diff
@@ -196,24 +193,19 @@ func (b *BM25Searcher) wand(
...
-					termResult, docIndices, err := b.createTerm(N, filterDocIds, queryTerms[j], propNames,
-						propertyBoosts, duplicateBoosts[j], params.AdditionalExplanations)
-					if err != nil {
-						return err
-					}
-					results[k] = termResult
-					indices[k] = docIndices
-					return nil
-				})
+				termResult, docIndices, err := b.createTerm(N, filterDocIds, queryTerms[j], propNames,
```
Comment (antas-marcin) [inline adapters/repos/db/inverted/bm25_searcher.go:196]: why also here `createTerm` method is run without a goroutine?

File: adapters/repos/db/lsmkv/segment_collection_strategy.go [hunk 3]
```diff
@@ -26,9 +26,11 @@ func (s *segment) getCollection(key []byte) ([]value, error) {
...
 	}
 
-	if !s.bloomFilter.Test(key) {
-		return nil, lsmkv.NotFound
-	}
+	/*
+		if !s.bloomFilter.Test(key) {
+			return nil, lsmkv.NotFound
+		}
+	*/
```
Comment (antas-marcin) [inline adapters/repos/db/lsmkv/segment_collection_strategy.go:33]: why it is commented?

File: test/acceptance_with_go_client/filters_tests/where_test.go [hunk 4]
```diff
@@ -12,12 +12,13 @@
...
 
 import (
-	acceptance_with_go_client "acceptance_tests_with_client"
 	"context"
 	"fmt"
 	"testing"
 	"time"
 
+	acceptance_with_go_client "acceptance_tests_with_client"
+
```
Comment (antas-marcin) [inline test/acceptance_with_go_client/filters_tests/where_test.go:21]: why did this import move? my formatter doesn't change this order

File: tools/dev/run_dev_server.sh [hunk 5]
```diff
@@ -16,11 +16,11 @@ export ORIGIN=${ORIGIN:-"http://localhost:8080"}
 export QUERY_DEFAULTS_LIMIT=${QUERY_DEFAULTS_LIMIT:-"20"}
 export QUERY_MAXIMUM_RESULTS=${QUERY_MAXIMUM_RESULTS:-"10000"}
 export TRACK_VECTOR_DIMENSIONS=true
-export CLUSTER_HOSTNAME=${CLUSTER_HOSTNAME:-"node1"}
+export CLUSTER_HOSTNAME=${CLUSTER_HOSTNAME:-"weaviate-0"}
```
Comment (antas-marcin) [inline tools/dev/run_dev_server.sh:19]: why did you change the default node name to `weaviate-0`?

File: tools/dev/run_dev_server.sh [hunk 6]
```diff
@@ -16,11 +16,11 @@ export ORIGIN=${ORIGIN:-"http://localhost:8080"}
...
 
 function go_run() {
   GIT_HASH=$(git rev-parse --short HEAD)
-  go run -ldflags "-X github.com/weaviate/weaviate/usecases/config.GitHash=$GIT_HASH" "$@"
+  go run -race -ldflags "-X github.com/weaviate/weaviate/usecases/config.GitHash=$GIT_HASH" "$@"
```
Comment (antas-marcin) [inline tools/dev/run_dev_server.sh:23]: is `-race` here necessary?

File: usecases/backup/scheduler.go [hunk 7]
```diff
@@ -261,7 +261,7 @@ func (s *Scheduler) validateRestoreRequest(ctx context.Context, store coordStore
 	if err != nil {
 		notFoundErr := backup.ErrNotFound{}
 		if errors.As(err, &notFoundErr) {
-			return nil, fmt.Errorf("%w: %q", errMetaNotFound, destPath)
+			return nil, fmt.Errorf("validateRestoreRequest: %w: '%v' (%v)", errMetaNotFound, destPath, err)
```
Comment (antas-marcin) [inline usecases/backup/scheduler.go:264]: please use `%w` for error wrapping

File: cmd/bucket-inspector/main.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/weaviate/weaviate/entities/cyclemanager"
+	"github.com/weaviate/weaviate/usecases/config"
+)
+
+func dumpBucket(storeDir, propName string) {
+	kvstore, err := lsmkv.New(storeDir, storeDir, &logrus.Logger{}, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("propName: %s\n", propName)
```
[the remark this answers is not in the document]
Comment (donomii) [reply cmd/bucket-inspector/main.go:32]: why?

Have a look at the file name

File: test/acceptance_with_go_client/filters_tests/where_test.go (hunk 4 above)
Comment (donomii) [reply test/acceptance_with_go_client/filters_tests/where_test.go:21]: It's correct go formatting

https://google.github.io/styleguide/go/decisions.html#import-grouping

File: adapters/repos/db/lsmkv/segment_collection_strategy.go (hunk 3 above)
Comment (donomii) [reply adapters/repos/db/lsmkv/segment_collection_strategy.go]: debugging

File: test/acceptance_with_go_client/filters_tests/where_test.go (hunk 4 above)
Comment (antas-marcin) [reply test/acceptance_with_go_client/filters_tests/where_test.go:21]: Which linter do you use (we agreed to use `golangci-lint` as a linter)? your import refactor makes my linter to mark the code in yellow saying that the file is not gofumted

File: test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go [hunk 9]
```diff
@@ -12,12 +12,13 @@
...
 
 import (
-	"acceptance_tests_with_client/fixtures"
 	"context"
 	"fmt"
 	"testing"
 	"time"
 
+	"acceptance_tests_with_client/fixtures"
+
```
Comment (antas-marcin) [inline test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go:21]: this line is now marked as not `gofumted` in my workspace, do you use `golangci-lint` tool as a linter?

File: usecases/backup/backupper.go [hunk 10]
```diff
@@ -103,7 +103,7 @@ func (b *backupper) OnStatus(ctx context.Context, req *StatusRequest) (reqStat,
 	meta, err := store.Meta(ctx, req.ID, false)
 	if err != nil {
 		path := fmt.Sprintf("%s/%s", req.ID, BackupFile)
-		return reqStat{}, fmt.Errorf("%w: %q: %v", errMetaNotFound, path, err)
+		return reqStat{}, fmt.Errorf("Backupper OnStatus: %w: %q: %v", errMetaNotFound, path, err)
```
Comment (redouan-rhazouani) [inline usecases/backup/backupper.go:106]: I would recommend against altering the error message. There's no need to incorporate debug information, such as the function name, in a message that is returned to public API users.

The principle thing applies to the changes below.

**Review by redouan-rhazouani — DISMISSED** [review]

File: usecases/backup/backupper.go (hunk 10 above)
Comment (donomii) [reply usecases/backup/backupper.go:106]: There is most definitely a need for altering the error message.  There are five separate places where this error message appears, and the error messages are completely indistinguishable.  Since weaviate strips all location information from the errors, there is no way to tell them apart.

Good, informative error messages are vital to good quality engineering, so there is a need to incorporate this information, and so these changes are required.

File: test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go (hunk 9 above)
Comment (donomii) [reply test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go:21]: Of course.

> jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)> golangci-lint run                                                                                                                                   (base) 
jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)> git diff                                                                                                                                            (base) 
jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)>

File: test/acceptance_with_go_client/filters_tests/where_test.go (hunk 4 above)
Comment (donomii) [reply test/acceptance_with_go_client/filters_tests/where_test.go:21]: I am using golangci-lint as the linter

> jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)> golangci-lint run                                                                                                                                   (base) 
jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)> git diff                                                                                                                                            (base) 
jeremyprice@Jeremys-MacBook-Pro ~/g/weaviate (weaviate-core-refactor)>

File: adapters/repos/db/inverted/bm25_searcher.go [hunk 11]
```diff
@@ -38,6 +37,7 @@ import (
 	"github.com/weaviate/weaviate/entities/schema"
 	"github.com/weaviate/weaviate/entities/searchparams"
 	"github.com/weaviate/weaviate/entities/storobj"
+	"golang.org/x/sync/errgroup"
```
Comment (antas-marcin) [inline adapters/repos/db/inverted/bm25_searcher.go:40]: can this file be reverted to it's original form? those changes introduced here are only empty spaces and a change in import statement. Those changes made here doesn't bring anything new to the project.

File: test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go (hunk 9 above)
Comment (antas-marcin) [reply test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go:21]:
```
weaviate/test/acceptance_with_go_client/multi_tenancy_tests (weaviate-core-refactor)$ golangci-lint run
activation_deactivation_test.go:14: File is not `gofumpt`-ed (gofumpt)
import (
activation_deactivation_test.go:19: File is not `gofumpt`-ed (gofumpt)

	"acceptance_tests_with_client/fixtures"
batch_reference_test.go:14: File is not `gofumpt`-ed (gofumpt)
import (
```

Comment (donomii) [reply test/acceptance_with_go_client/multi_tenancy_tests/activation_deactivation_test.go:21]: What version are you using?  It's passing all the github actions tests, and they include gofumpt

File: usecases/backup/backupper.go (hunk 10 above)
Comment (redouan-rhazouani) [reply usecases/backup/backupper.go:106]: > There is most definitely a need for altering the error message. There are five separate places where this error message appears, and the error messages are completely indistinguishable. Since weaviate strips all location information from the errors, there is no way to tell them apart.

First, it's worth mentioning that the current message is intended for the public API.

> Good, informative error messages are vital to good quality engineering, so there is a need to incorporate this information, and so these changes are required.

However, while I certainly value the importance of informative error messages, it seems that the new message might not be the most suitable for a couple of reasons:

1. It appears to leak internal debugging information to public API users.
2. Its format resembles that of a logging message rather than one intended for the public AP

Having said that, I am not against changing the message to a more descriptive one if it can provide meaningful information to users without exposing internal details.

**Comment by donomii** [comment]
I agree it is nice to have simple messages for the public API, but can we
also have more detailed messages for debugging?

File: usecases/backup/backupper.go (hunk 10 above)
Comment (donomii) [reply usecases/backup/backupper.go:106]: Huh, replying to emails doesn't reply here.

I agree it would be nice to have simple messages for the public API, but is there a way we could also have good debugging messages for debugging?

Also
> It appears to leak internal debugging information to public API users.

we're open source, we don't have to worry about that

File: adapters/repos/db/repo.go [hunk 12]
```diff
@@ -76,6 +76,31 @@ type DB struct {
 	ratePerSecond       int
 }
 
+func (db *DB) GetSchemaGetter() schemaUC.SchemaGetter {
+	return db.schemaGetter
+}
+
+func (db *DB) GetSchema() schema.Schema {
+	return db.schemaGetter.GetSchemaSkipAuth()
```
Comment (parkerduckworth) [inline adapters/repos/db/repo.go:84]: Are we certain there is no harm in side-stepping `db.schemaGetter.GetSchemaSkipAuth()` here? I have seen elsewhere in the codebase that getting the schema is a protected action, and skipping auth is only accessible by some privileged packages, which are assumed to have already authenticated/authorized

File: cmd/weaviate-core/main.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+	"github.com/weaviate/weaviate/adapters/handlers/rest"
+	"github.com/weaviate/weaviate/entities/storobj"
+	"github.com/weaviate/weaviate/usecases/config"
+)
+
+func main() {
+	dataPath := flag.String("data-path", "./data", "path to the data directory")
+	origin := flag.String("origin", "http://localhost:8080", "listen address")
+	clusterHostname := flag.String("cluster-hostname", "node1", "cluster hostname")
+	defaultVectoriser := flag.String("default-vectoriser", "none", "default vectoriser")
```
Comment (parkerduckworth) [inline cmd/weaviate-core/main.go:30]: `vectoriser` is spelled `vectorizer` elsewhere in the codebase

Comment (donomii) [reply cmd/weaviate-core/main.go:30]: I don't accept your fancy new language innovations :P

File: adapters/repos/db/repo.go (hunk 12 above)
Comment (donomii) [reply adapters/repos/db/repo.go:84]: The other examples don't appear to be any more protected than this one, e.g. c9213605e70346188755d4890cc72b0f5cecef2c (ListClasses), is in the same package, and does not appear to have protection.

I also don't see the point of protecting the schema in preference to the data, it's not like we keep passwords in the schema.  

It looks like the API on schemaGetter is not fitting, since we have authentication on the network calls, not on the database itself.

Can you point out the potential vulnerability here?  It's not like users can call arbitrary functions.

**Comment by donomii** [comment]
obsolete
